lib/gc/Transforms/GPU/AllocsToSLM.cpp

```diff
@@ -62,22 +65,88 @@ struct ConvertAlloc : public OpRewritePattern<memref::AllocOp> {
                                          "Only support allocs in GPU regions");
     }
 
+    auto launchOp = allocOp->getParentOfType<gpu::LaunchOp>();
+
+    auto xSz = dyn_cast<arith::ConstantIndexOp>(
+        launchOp.getBlockSizeX().getDefiningOp());
+    auto ySz = dyn_cast<arith::ConstantIndexOp>(
+        launchOp.getBlockSizeY().getDefiningOp());
+    auto zSz = dyn_cast<arith::ConstantIndexOp>(
+        launchOp.getBlockSizeZ().getDefiningOp());
+
+    if (!xSz || !ySz || !zSz)
+      return rewriter.notifyMatchFailure(
+          allocOp, "Only support constant block sizes for now");
+
+    int64_t xI = xSz.value();
+    int64_t yI = ySz.value();
+    int64_t zI = zSz.value();
+
+    if (zI != 1)
+      return rewriter.notifyMatchFailure(
+          allocOp, "Only support 2D shared memory for now");
+
+    int64_t totalWorkGroupSize = xI * yI * zI;
+
     Value memref = allocOp->getResult(0);
+
     MemRefType originalMemRefType = cast<MemRefType>(memref.getType());
 
+    // Scale the allocation size by the number of threads in the work-group
+    int64_t newX = originalMemRefType.getShape()[0] * xI;
+    int64_t newY = originalMemRefType.getShape()[1] * yI;
+
+    SmallVector<int64_t> newShape = {newX, newY};
+
     IntegerAttr sharedAddressSpace =
         IntegerAttr::get(rewriter.getIntegerType(64),
                          static_cast<int64_t>(gpu::AddressSpace::Private));
 
-    // Create a new MemRefType with the desired address space
-    MemRefType newMemRefType = MemRefType::get(
-        originalMemRefType.getShape(), originalMemRefType.getElementType(),
-        originalMemRefType.getLayout(), sharedAddressSpace);
-
-    Value newMemRef = rewriter.create<memref::AllocOp>(
-        allocOp.getLoc(), newMemRefType, allocOp.getOperands());
-
-    memref.replaceAllUsesWith(newMemRef);
+    MemRefType newRootMemRefType =
+        MemRefType::get(newShape, originalMemRefType.getElementType(),
+                        originalMemRefType.getLayout(), sharedAddressSpace);
+
+    Value newRootMemRef =
+        rewriter
+            .create<memref::AllocOp>(allocOp.getLoc(), newRootMemRefType,
+                                     allocOp.getOperands())
+            .getResult();
+
+    // Compute the offsets in SLM chunk for the current thread
+    auto oneConst =
+        rewriter.create<arith::ConstantIndexOp>(allocOp.getLoc(), 1);
+    auto origXConst = rewriter.create<arith::ConstantIndexOp>(
+        allocOp.getLoc(), originalMemRefType.getShape()[0]);
+    auto origYConst = rewriter.create<arith::ConstantIndexOp>(
+        allocOp.getLoc(), originalMemRefType.getShape()[1]);
+
+    auto threadIds = launchOp.getThreadIds();
+
+    auto offX =
+        rewriter
+            .create<arith::MulIOp>(allocOp.getLoc(), threadIds.x, origXConst)
```

Comment:
Are all the values constants? If yes, the result could be precalculated instead of using MulIOp.

Reply:
no, `threadIds.x` is a thread's index along X axis, it's not a constant